File: solution/51/solution.py
import copy
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        result = []
        item = []
        onelines = ['.'*y + 'Q' + '.'*(n-y-1) for y in xrange(n)]
        marks = [[0]*n for _ in xrange(n)]
        self.complate(0, n, item, result, marks, onelines)
        return result

    def updatemark(self, x, y, marks):
        marks[x][y] = 1
        markx = [0, 0, 1, 1, 1, -1, -1, -1];
        marky = [-1, 1, 0, -1, 1, 0, -1, 1];
        n = len(marks)
        markn = max(n-x, x, n-y, y)
        for i in xrange(1, markn):
            for j in xrange(8):
                newx = x + i * markx[j]
                newy = y + i * marky[j]
                if newx >=0 and newx < len(marks) and newy >= 0 and newy < len(marks):
                    marks[newx][newy] = 1

    def complate(self, x, n, item, result, marks, onelines):
        if x == n:
            result.append(copy.copy(item))
            return
        for y in xrange(n):
            if marks[x][y] == 1:
                continue
            item.append(onelines[y])
            copymark = copy.deepcopy(marks)
            self.updatemark(x, y, marks)
            self.complate(x+1, n, item, result, marks, onelines)
            marks = copymark
            item.pop()

s = Solution()
print len(s.solveNQueens(9))

File: solution/51/solution.cpp
#include <stdio.h>
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::vector<std::string> > solveNQueens(int n) {
        std::vector<std::vector<std::string> > result;
        std::vector<std::string> item;
        std::vector<std::string> nowboard;
        for(int i=0;i< n;i++){
            std::string s;
            for(int x=0;x< n;x++){
                s += '.';
            }
            nowboard.push_back(s);
            item.push_back(s);
        }
        complateNQueue(0, item, result, nowboard);
        return result;
    }
    void updateboard(std::vector<std::string>& nowboard, int x, int y){
        nowboard[x][y] = '1';
        static const int markx[] = {0, 0, 1, 1, 1, -1, -1, -1};
        static const int marky[] = {-1, 1, 0, -1, 1, 0, -1, 1};
        for(int i=0;i<nowboard.size();i++){
            for(int j=0;j<8;j++){
                int newx = x + markx[j]*i;
                int newy = y + marky[j]*i;
                if(newx>=0&&newx<nowboard.size()&&newy>=0&&newy<nowboard.size())nowboard[newx][newy]='1';
            }
        }
    }
    void complateNQueue(int x, std::vector<std::string>& item, std::vector<std::vector<std::string> >& result, std::vector<std::string>& nowboard){
        if(x== nowboard.size()){result.push_back(item);return;}
        for(int y=0;y<nowboard.size();y++){
            if(nowboard[x][y]=='1')continue;
            std::vector<std::string> copyboard = nowboard;
            item[x][y] = 'Q';
            updateboard(nowboard, x, y);
            complateNQueue(x+1, item, result, nowboard);
            nowboard = copyboard;
            item[x][y]='.';
        }
    }
};


int main(){
    Solution s;
    s.solveNQueens(1);
}